### PR TITLE
wasm: fix race condition on listener info in contexts.

### PR DESCRIPTION
**Pull request by mathetake**
This is the follow up on #16795.

## Bug description

If the same plugin/rootcontext is shared across multiple listeners, then there is a race in listener information.

Before  #16795,  this was not a problem  for stream contexts since given listener information is from *where* it is created (`Plugin: plugin_` member created in the factory instance) https://github.com/envoyproxy/envoy/pull/16795/files#diff-90e133f83d91e5c22d1f7c7e8c0cf8017536a23765978bf81c7efd0482be0d81L43
But after that being merged, the `Plugin` passed to stream context constructor is the one from `PluginHandle::plugin()` which is potentially shared across multiple listeners.

For root contexts, this has been a problem since the beginning because the root context has been already potentially shared across multiple listeners, and pointing to the listener's info which is created first.

## Change in this PR

Combined with https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/178, we explicitly include listener information into the plugin key so that if two plugins have different info to each other, different plugin instances will be created.

## Note

- Maybe we could drop the support for sharing 1 Plugin across multiple listeners (or filters), then this race condition would be resolved in a different way. I think it is worth discussion.
- This race is originally reported by Istio https://github.com/istio/proxy/pull/3395#issuecomment-873477432

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: source/extensions/filters/http/wasm/wasm_filter.h
Comment (lambdai) [inline source/extensions/filters/http/wasm/wasm_filter.h:51]: both const? Here as http filter and below as network filter.

File: source/extensions/common/wasm/context.h
Comment (lambdai) [inline source/extensions/common/wasm/context.h:478]: Regarding the comment, does network filter allow to use it as well? in Non-wasm it is allowed. And you do set  direction_ in network filter config 
I could be wrong since I don't have too much contexts on wasm :)

**Review by lambdai — COMMENTED** [review]
Lifetime of listener metadata LGTM

**Comment by kyessenov** [comment]
@PiotrSikora What does Wasm spec say here? There are some significant repercussions on how we share roots since caches are generally stored there. Stats filter in istio, for example, keeps track of request reports in the root and (theoretically) benefits from sharing. I remember that the big change was making sure that listener reloads create new roots to avoid immediate config application. If we have multiple listeners reloading, how do we make sharing work?

File: source/extensions/common/wasm/context.h
Comment (mathetake) [reply source/extensions/common/wasm/context.h:478]: yeas it does. In Wasm, "stream context"  means each Http or Tcp/Network stream, fyi:)

File: source/extensions/filters/http/wasm/wasm_filter.h
Comment (mathetake) [reply source/extensions/filters/http/wasm/wasm_filter.h]: done. Thanks! https://github.com/envoyproxy/envoy/pull/17243/commits/00a45f4ecde4e7cdca3173325b330c219b99be89

File: test/extensions/filters/http/wasm/test_data/test_cpp.cc
Comment (mathetake) [inline test/extensions/filters/http/wasm/test_data/test_cpp.cc:75]: I just realized this also could be a race, but I would do in another PR.

**Comment by PiotrSikora** [comment]
@mathetake thanks for the fast turnaround on the fix and detailed description of the changes in this PR, that's really helpful!

While this PR looks fine code-wise, I'd prefer if we didn't introduce changes breaking existing Wasm extensions. Making listener info unavailable in the root context also doesn't seem like a desireable outcome.

The original intent was always to create a separate root context for each unique configuration, with multiple Envoy filters configured _exactly_ _the_ _same_ _way_ being backed by a single root context inside the WasmVM. Currently, "unique configuration" is keyed only on `root_id` and plugin configuration, but it seems that the key should also include listener metadata and traffic direction. Such change should only slightly increase the number of root contexts without breaking anything, and it's technically more of a bugfix than a behavioral change.

Alternatively, we could change the mapping and create a separate root context for each configured Envoy filter, which I think could be easier to reason about to the users, but as @kyessenov pointed out, we'd lose the benefit of sharing root context across multiple (potentially thousands) identially configured Envoy filters, which might be quite costly at scale.

**Comment by mathetake** [comment]
> it seems that the key should also include listener metadata and traffic direction. Such change should only slightly increase the number of root contexts without breaking anything, and it's technically more of a bugfix than a behavioral change.

OK that would work and make sense to me. Let me rework this PR and create the corresponding one to C++ host. (Sorry in advance I won't be available until tomorrow morning in UTC+9 so don't expect me to turnaround in the next 24 hours)

**Comment by mathetake** [comment]
OK reworked this PR, and created the necessary change in C++ host: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/178

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:979]: /wait until this is merged in Proxy-Wasm C++ Host repo.

**Review by PiotrSikora — COMMENTED** [review]
LGTM, but could you verify that this fixes the issue in Istio Proxy before we merge it? Thanks!

**Comment by mathetake** [comment]
verified that the original issue Istio faced disappeared!

**Comment by mathetake** [comment]
ping @moderation for dep:)

File: source/extensions/filters/http/wasm/wasm_filter.h
Comment (PiotrSikora) [inline source/extensions/filters/http/wasm/wasm_filter.h:48]: Would it make sense to leave this and add an assert to verify that `plugin == handle->plugin()`?

**Review by PiotrSikora — APPROVED** [review]
Thanks!

File: source/extensions/filters/http/wasm/wasm_filter.h
Comment (mathetake) [reply source/extensions/filters/http/wasm/wasm_filter.h:48]: in order to do that, we have to define == operator on PluginBase and Plugin classes because `plugin_` is the "new" shared_ptr than the one stored in PluginHandle. I think that would make complicates the implementation. wdyt?

Comment (PiotrSikora) [reply source/extensions/filters/http/wasm/wasm_filter.h:48]: OK, let's skip it, then.

**Pull request merged**
